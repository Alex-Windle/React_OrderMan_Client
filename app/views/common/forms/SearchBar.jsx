/**
 * GENERAL NOTES
 * @author TalkRise <admin@talkrise.com>
 */


// Module imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { debounce } from 'lodash';


/**
 * @class SearchBar
 * @description Represents an input that responds and debounces onChange events
 *
 */
export default class SearchBar extends Component {
  constructor(props) {
    super(props);

    /**
     * Use lodash debounce to debounce the handleSearch function
     */
    this.debouncedHandleSearch = debounce(this.props.handleSearch, 500);


    /**
     * @function _handleSearch
     * @description Debounces the change event and calls handler passed into component
     * @param event
     */
    this._handleSearch = (event) => {
      /**
       * When you debounce React events, React tries to reuse synthetic events for you for
       * efficiency purposes. To persist the original event and get it's data, you must call
       * 'event.persist()'.
       * (http://stackoverflow.com/questions/35435074/using-debouncer-with-react-event
       */
      event.persist();
      const value = event.target.value;
      this.debouncedHandleSearch(value);
    };
  }


  render() {
    const { disabled, placeholder } = this.props;
    return (
      <div className="form-group">
        <input
          autoComplete="off"
          className="form-control"
          disabled={disabled}
          onChange={this._handleSearch}
          placeholder={placeholder}
          type={'text'}
        />
      </div>
    );
  }
}


SearchBar.propTypes = {
  disabled: PropTypes.bool,
  handleSearch: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
};


SearchBar.defaultProps = {
  disabled: true,
  placeholder: 'Search',
};
