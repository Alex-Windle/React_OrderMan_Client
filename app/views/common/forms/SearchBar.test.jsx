/**
 * GENERAL NOTES
 * @author TalkRise <admin@talkrise.com>
 */


// Module imports
import React from 'react';
import renderer from 'react-test-renderer';


// Component imports
import SearchBar from './SearchBar';


describe('SearchBar component', () => {
  it('matches snapshot', () => {
    const tree = renderer.create(
      <SearchBar
        handleSearch={jest.fn()}
        disabled={false}
        placeholder="Placeholder"
      />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
