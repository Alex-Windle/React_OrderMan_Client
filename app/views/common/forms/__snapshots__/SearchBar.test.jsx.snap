// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar component matches snapshot 1`] = `
<div
  className="form-group"
>
  <input
    autoComplete="off"
    className="form-control"
    disabled={false}
    onChange={[Function]}
    placeholder="Placeholder"
    type="text"
  />
</div>
`;
